### ISSUE TITLE
Impossible to handle non root directory for project

### ISSUE THREAD

**Issue by C0DK** [heroku/heroku-buildpack-nodejs#787]
This is a general issue with Heroku buildpacks but i am not sure how to get around it.
We recently switched to a more separated structure; the root directory has 4 folders:
`backend`, `webapp`, `app`, `assets`, and i am able to make it work fine locally but ended up having to use docker to even be able to host it on your services, but now i have the same issue with heroku tests, where i'm not able to get our CI to run because the buildpack cannot find the package.json folder. For the heroku stack we want it to use `app/package.json`, however we need to have two different `package.json` for our web app and our phone app, so it sorta sucks having one of them in the project root folder.

Personally I like have them all in repository to make it sure all things have the same version. Is the only solution to split up or application into different applications? That poses new challenges - IMO - right now we could easily just serve the compiled webapp via our backend, but that would be difficult if they are in different repositories all together.

What is recommended? also is would it be looked into to be able to pass a path to various buildpacks?

**Comment by danielleadams** [comment]
Hi @C0DK, Is this a duplicate of https://github.com/heroku/heroku-buildpack-nodejs/issues/385? Monorepository support will be in NPM 7, so we're waiting on that release to do work here. (See workspaces under "Coming Attractions" here: https://blog.npmjs.org/post/617484925547986944/npm-v7-series-introduction)

**Comment by laguiz** [comment]
Hi,
What about if the backend is not node? Would be interesting to be able to set the folder where to run the build pack. In my case I need to compile an Angular app in 'front' folder and copy to backend public asset folder.

Idea is to stay in Heroku standards to leverage pipelines etc.

**Comment by danielleadams** [comment]
@laguiz I understand the use case, but what would be the downside of having the `package.json` at the root though?

If you are using a monorepo structure, you will have more options by having the `package.json` at root and treating one or more subdirectories (ie. `frontend`, `ui`, `mobile`, `package-a`, etc) as a workspace, which has benefits for dependency management.

**Comment by laguiz** [comment]
@danielleadams 
Yes that's what I did. Well kind of...

For now I use 4 buildpacks :
1. Node with very basic package.json at root. This basically just make node/npm available for step 2
2. Run simple sh file to build (npm install and ng/build) and copy compiled assets from front/dist to my backend public folder. With sh script I can easly "move" to my front folder to make it "my node workspace".
3. Convert project to make my backend the root folder (remove all other folders/files that are not needed after assets are compiled).
4. Java Play sbt based on root folder

It works pretty well so far. If I stay like this, I will need to improve caching of node_modules myself I guess.

If I understand well, you say that it may be possible to 'merge' step 2 (manual sh script) into step 1 to run the build with native nodejs buildpack. If so, could you point a documentation to 'set' the root path (workspace?) of the build in a sub folder?

Sorry about noob question, I am still new on the concept of buildpack.

Thank you for your great work.

**Comment by C0DK** [comment]
> @laguiz I understand the use case, but what would be the downside of having the `package.json` at the root though?

We have a mono rep, with `backend` `webapp` and `app` - having the package.json in root would mean that the webapp and app are required to have the same dependencies. This is naturally not the case. Also it is just less clean, in general, to have a package.json outside the scope in which it is relevant IMO. I like having a clean root directory. Unless that is what we could do with workspaces?

**Comment by C0DK** [comment]
But yes it is basically a duplicate, and should probably be closed.

**Comment by danielleadams** [comment]
@C0DK workspaces manages dependencies for mono repos. It means you can have different dependencies that are specified in `package.json` files under child directories. Please see here:

- yarn: https://classic.yarnpkg.com/en/docs/workspaces/
- npm plans to release workspaces with v7: https://blog.npmjs.org/post/617484925547986944/npm-v7-series-introduction
- npm RFC for workspaces: https://github.com/npm/rfcs/pull/103/files
- lerna is a package that you can manage monorepos: https://github.com/lerna/lerna

Sure, it adds an additional file to the root directory, which could be annoying, but this is also the convention that all of the above projects have settled on that provides the most clear configuration for mono repos.

**Comment by danielleadams** [comment]
hey @laguiz, not noob questions at all. workspaces is a concept that the package managers (yarn, pnpm, and now npm) have adopted to handle any monorepo file structure. I'd like to write some documentation on creating monorepos for Heroku apps once NPM 7 has been released (which will handle workspaces). It will take care of you wanting to cache at the subdirectory level.

In the meantime, have you tried custom caching directories? It's supported on Heroku: https://devcenter.heroku.com/articles/nodejs-support#custom-caching

**Comment by laguiz** [comment]
@danielleadams Great got it. I think Heroku custom cache will fill my needs 👍 I will give it a try. Thank you for the link.

**Comment by danielleadams** [comment]
Closing because it's a duplicate! Please see https://github.com/heroku/heroku-buildpack-nodejs/issues/385
